### PR TITLE
Hotfix/1310 immutable body list

### LINKED ISSUE #1310: New bug with Star scanned script from `feature/scan-overhaul2`
## EDDI version in which issue is found
Branch `feature/scan-overhaul2`

## VoiceAttack version in which issue is found (as applicable)
Standalone

## Steps to reproduce
1. Jumped into `Eactaiky AH-B c27-13` and honked the FSS.

## Expected
Normal scan report.

## Observed
EDDI said:
```
There is a problem with the: Star scanned script. Exception has been thrown by the target of an invocation.
```

## Investigation
L:og shows:
```
2019-05-22T21:32:56 [Warning] ScriptResolver:resolveScript Failed to resolve the: Star scanned script. System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Collections.Generic.List`1.Enumerator.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__23`3.MoveNext()
   at System.Linq.Enumerable.<DistinctIterator>d__64`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at EddiDataDefinitions.StarSystem.get_materialEdNames() in C:\dev\EDDI\DataDefinitions\StarSystem.cs:line 133
   at EddiDataDefinitions.StarSystem.get_isgold() in C:\dev\EDDI\DataDefinitions\StarSystem.cs:line 108
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Cottle.Values.ReflectionValue.MemberReader.<>c__DisplayClass6_0.<.ctor>b__0(Object s)
   at Cottle.Values.ReflectionValue.MemberReader.Extract(Object source)
   at Cottle.Values.ReflectionValue.Resolve()
   at Cottle.Values.ResolveValue.Acquire()
   at Cottle.Values.ResolveValue.get_Fields()
   at Cottle.Documents.Simple.Evaluators.AccessEvaluator.Evaluate(IStore store, TextWriter output)
   at Cottle.Documents.Simple.Evaluators.InvokeEvaluator.Evaluate(IStore store, TextWriter output)
   at Cottle.Documents.Simple.Nodes.EchoNode.Render(IStore store, TextWriter output, Value& result)
   at Cottle.Documents.Simple.Nodes.CompositeNode.Render(IStore store, TextWriter output, Value& result)
   at Cottle.Documents.SimpleDocument.Render(IStore store, TextWriter writer)
   at Cottle.Documents.AbstractDocument.Render(IStore store)
   at EddiSpeechResponder.ScriptResolver.resolveScript(String script, BuiltinStore store, Boolean master, Script scriptObject) in C:\dev\EDDI\SpeechResponder\ScriptResolver.cs:line 88 
```

### PR DESCRIPTION
**Pull request by richardbuckle**
I came to realise that as long as `StarSystem` has  a mutable `List<Body>` property, it cannot be made thread-safe because some other thread can always add to or update the list at any time. So now that property is an `ImmutableList<Body>` and the `StarSystem` class exposes methods to add and update bodies. This is actually much nicer.

Weirdly, .NET's immutable container classes were never added to the solution -- that's fixed now of course.

Fixes #1310

### CODE REVIEW

File: Tests/DataDefinitionTests.cs [hunk 1]
```diff
@@ -108,26 +108,40 @@ public void TestOcellusStationModel()
...
             {
                 bodyname = "Jumponium Haven",
                 materials = materials
             };
 
+            // Add a body with green materials and re-test
+            StarSystem starSystem = new StarSystem() { systemname = "testSystem" };
+            starSystem.AddBody(jumponiumBody);
+            Assert.IsTrue(starSystem.isgreen, "green system should be green");
+            Assert.IsFalse(starSystem.isgold, "green system should be gold");
```
Comment (Tkael) [inline Tests/DataDefinitionTests.cs]: should not be?

Comment (richardbuckle) [reply Tests/DataDefinitionTests.cs]: You got me! Feel free to push a fix.

Assertions w/o comments suck; so do assertions with incorrect comments. Same as all other code.

**Comment by richardbuckle** [comment]
For the record, my Gist on why we needed to do this: https://gist.github.com/richardbuckle/e6064ae2a2d2c9be1f303231d669a289

File: EDDI/EDDI.cs [hunk 2]
```diff
@@ -1927,71 +1927,30 @@ private bool eventNearSurface(NearSurfaceEvent theEvent)
...
-                    CurrentStarSystem.Reserve = theEvent.body.reserveLevel;
-                }
+            CurrentStarSystem.AddOrUpdateBody(theEvent.body);
 
-                Logging.Debug("Saving data for scanned body " + theEvent.bodyname);
-                StarSystemSqLiteRepository.Instance.SaveStarSystem(CurrentStarSystem);
+            // Update the system reserve level, when appropriate
+            if (theEvent.body.reserveLevel != ReserveLevel.None)
+            {
+                CurrentStarSystem.Reserve = theEvent.body.reserveLevel;
```
Comment (Tkael) [inline EDDI/EDDI.cs]: We could automate this like we do system materials, so that it auto-updates as we add bodies.

Comment (richardbuckle) [reply EDDI/EDDI.cs]: Good call.

**Pull request merged**